smac_planner/include/smac_planner/costmap_downsampler.hpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2020, Carlos Luis
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License. Reserved.
+
+#ifndef SMAC_PLANNER__DOWNSAMPLER_HPP_
+#define SMAC_PLANNER__DOWNSAMPLER_HPP_
+
+#include <algorithm>
+#include "nav2_costmap_2d/costmap_2d_ros.hpp"
+#include "smac_planner/constants.hpp"
+
+namespace smac_planner
+{
+
+/**
+ * @class smac_planner::CostmapDownsampler
+ * @brief A costmap downsampler for more efficient path planning
+ */
+class CostmapDownsampler
+{
+public:
+  /**
+   * @brief A constructor for CostmapDownsampler
+   */
+  explicit CostmapDownsampler()
+  : _sampling_factor(0)
+  {
+  }
+
+  /**
+   * @brief A destructor for CostmapDownsampler
+   */
+  ~CostmapDownsampler()
+  {
+  }
+
+  /**
+   * @brief Downsample the given costmap to the closest approximation of the target resolution
+   * @param original_costmap The costmap we want to downsample
+   * @param target_resolution The desired size of a cell in the new costmap, in meters
+   * @return A pointer to the newly created downsampled costmap
+   * Note: despite returning a pointer, this class keeps ownership of the resource it created.
+   */
+  nav2_costmap_2d::Costmap2D * downsample(
+    nav2_costmap_2d::Costmap2D * original_costmap,
+    const float & target_resolution)
+  {
+    _original_size_x = original_costmap->getSizeInCellsX();
+    _original_size_y = original_costmap->getSizeInCellsY();
+    _original_costmap = original_costmap;
+    _sampling_factor = round(target_resolution / original_costmap->getResolution());
```

Comment:
This rounding makes it not actually do what the user asked - I think rather, we shold be working with a resolution multiplier so there's no rounding.

Reply:
It would also let you remove the `ceil`s below

Reply:
Right, I'll change it so that the param that we're passing is actually the sampling factor (what you call the resolution mulltiplier)

Reply:
With the change I can remove the `round` call, but not the `ceil`s. Those are necessary to make sure that the downsampled costmap covers completely the area of the original costmap. If we don't round up then there could be cells not included in the top and right borders of the map.